File: modules/vescape-core/ios/auth/NativeAuthCoordinator.swift
import Foundation

/// Device Token lifecycle: verify a freshly exchanged credential, store it, revoke it. The HTTP
/// boundary itself belongs to `VescapeApi` so every native caller shares one credential and one 401
/// policy.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/auth/NativeAuthCoordinator.kt
final class NativeAuthCoordinator {
  static let shared = NativeAuthCoordinator()
  private let store = DeviceCredentialStore.shared

  private static let accountPath = "/api/account"
  private static let revokePath = "/api/auth/device-tokens/current"

  func stateMap() throws -> [String: Any?] {
    let credential = try store.read()
    return [
      "state": store.state(credential: credential).rawValue,
      "accountId": credential?.accountId,
      "expiresAt": credential?.expiresAt,
    ]
  }

  /// Verifies the exchanged token against the Account it claims before storing it. The token is not
  /// in the store yet, so the call carries it explicitly.
  func provision(
    serverUrl: String,
    token: String,
    accountId: String
  ) async throws -> [String: Any?] {
    let origin = serverUrl.hasSuffix("/") ? String(serverUrl.dropLast()) : serverUrl
    let result: ApiResult<String> = await VescapeApi.forOrigin(origin).request(
      .get,
      path: Self.accountPath,
      auth: .bearer(token)
    ) { body in
      guard let data = body.data(using: .utf8),
            let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
            let id = json["id"] as? String
      else { throw NSError(domain: "NativeAuth", code: -3) }
      return id
    }

    switch result {
    case .ok(let returnedId):
      guard returnedId == accountId else { throw NSError(domain: "NativeAuth", code: -4) }
    // `VescapeApi` already rejected the stored credential and refreshed App Status.
    case .unauthorized:
      throw NSError(domain: "NativeAuth", code: 401)
    case .cancelled:
      throw CancellationError()
    default:
      throw NSError(domain: "NativeAuth", code: -5)
    }

    let credential = DeviceCredential(
      serverUrl: origin,
      token: token,
      accountId: accountId,
      expiresAt: nil
    )
    try store.write(credential)
    await MainActor.run {
      AppStatusCoordinator.shared.refresh()
    }
    return try stateMap()
  }

  /// Revokes server-side before the local copy goes away. A `401` means the server already considers
  /// it gone, which is the same end state.
  func revoke() async throws {
    guard let credential = try store.read() else { return }
    let result: ApiResult<Void> = await VescapeApi.forOrigin(credential.serverUrl).request(
      .delete,
      path: Self.revokePath,
      auth: .required
    ) { _ in () }

    switch result {
    case .ok, .unauthorized: break
    case .cancelled: throw CancellationError()
    default: throw NSError(domain: "NativeAuth", code: -6)
    }
    try store.clear()
  }

  func clear() throws { try store.clear() }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/auth/NativeAuthCoordinator.kt
package expo.modules.vescapecore.auth

import android.content.Context
import expo.modules.vescapecore.api.ApiResult
import expo.modules.vescapecore.api.AuthMode
import expo.modules.vescapecore.api.HttpMethod
import expo.modules.vescapecore.api.VescapeApi
import expo.modules.vescapecore.appstatus.AppStatusCoordinator
import org.json.JSONObject

/**
 * Device Token lifecycle: verify a freshly exchanged credential, store it, revoke it. The HTTP
 * boundary itself belongs to [VescapeApi] so every native caller shares one credential and one 401
 * policy.
 *
 * @parity /modules/vescape-core/ios/auth/NativeAuthCoordinator.swift
 */
class NativeAuthCoordinator(private val context: Context) {
  private val store = DeviceCredentialStore(context)

  fun stateMap(): Map<String, Any?> {
    val credential = store.read()
    return mapOf(
      "state" to store.state(credential).slug,
      "accountId" to credential?.accountId,
      "expiresAt" to credential?.expiresAt,
    )
  }

  /**
   * Verifies the exchanged token against the Account it claims before storing it. The token is not
   * in the store yet, so the call carries it explicitly.
   */
  suspend fun provision(
    serverUrl: String,
    token: String,
    accountId: String,
  ): Map<String, Any?> {
    val origin = serverUrl.trimEnd('/')
    val result = VescapeApi.forOrigin(context, origin).request(
      method = HttpMethod.GET,
      path = ACCOUNT_PATH,
      auth = AuthMode.Bearer(token),
    ) { body -> JSONObject(body).optString("id") }

    when (result) {
      is ApiResult.Ok ->
        check(result.value == accountId) { "Account verification mismatch" }
      // `VescapeApi` already rejected the stored credential and refreshed App Status.
      ApiResult.Unauthorized -> throw IllegalStateException("Device credential rejected")
      else -> throw IllegalStateException("Account verification failed ($result)")
    }

    store.write(DeviceCredential(origin, token, accountId, null))
    AppStatusCoordinator.get(context).refresh()
    return stateMap()
  }

  /**
   * Revokes server-side before the local copy goes away. A `401` means the server already considers
   * it gone, which is the same end state.
   */
  suspend fun revoke() {
    val credential = store.read() ?: return
    val result = VescapeApi.forOrigin(context, credential.serverUrl).request(
      method = HttpMethod.DELETE,
      path = REVOKE_PATH,
      auth = AuthMode.Required,
    ) { }

    when (result) {
      is ApiResult.Ok, ApiResult.Unauthorized -> Unit
      else -> throw IllegalStateException("Device credential revocation failed ($result)")
    }
    store.clear()
  }

  fun clear() = store.clear()

  companion object {
    private const val ACCOUNT_PATH = "/api/account"
    private const val REVOKE_PATH = "/api/auth/device-tokens/current"

    @Volatile private var instance: NativeAuthCoordinator? = null
    fun get(context: Context): NativeAuthCoordinator =
      instance ?: synchronized(this) {
        instance ?: NativeAuthCoordinator(context.applicationContext).also { instance = it }
      }
  }
}
